### PR TITLE
Add value type

### PR DESCRIPTION
**Pull request by amanmahajan7**
I will fix the following in a future PR
- Fix key type 
- Fix value type in Editors and Formatters. Need to finalize how to handle it internally

### CODE REVIEW

File: packages/react-data-grid/src/ColumnMetrics.ts [hunk 1]
```diff
@@ -114,8 +114,8 @@ function compareEachColumn<R>(prevColumns: ColumnList<R>, nextColumns: ColumnLis
 
   for (const key of keys) {
     if (!prevColumnsMap.has(key) || !nextColumnsMap.has(key)) return false;
-    const prevColumn = prevColumnsMap.get(key)!;
-    const nextColumn = nextColumnsMap.get(key)!;
+    const prevColumn = prevColumnsMap.get(key) as Column<R>;
```
Comment (amanmahajan7) [inline packages/react-data-grid/src/ColumnMetrics.ts]: `!` turns `Column<R>` to `NonNullable<Column<R>>` so had to explicitly cast it

File: packages/react-data-grid/src/HeaderCell.tsx [hunk 2]
```diff
@@ -3,15 +3,15 @@ import classNames from 'classnames';
...
 
-function SimpleCellRenderer<R>({ column, rowType }: { column: CalculatedColumn<R>; rowType: HeaderRowType }) {
+function SimpleCellRenderer<R>({ column, rowType }: HeaderRowProps<R>) {
   const headerText = rowType === HeaderRowType.HEADER ? column.name : '';
   return <div>{headerText}</div>;
 }
 
 interface Props<R> {
-  renderer?: CalculatedColumn<R>['headerRenderer'];
+  renderer?: React.ReactElement | React.ComponentType<HeaderRowProps<R>>;
```
Comment (amanmahajan7) [inline packages/react-data-grid/src/HeaderCell.tsx]: `CalculatedColumn<R>['headerRenderer']` was failing. Need to revisit it

File: packages/react-data-grid/src/common/types.ts [hunk 3]
```diff
@@ -7,11 +7,11 @@ export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
 
 export type SelectedRow<TRow> = TRow & { isSelected: boolean };
 
-export interface Column<TRow, TValue = unknown, TDependentValue = unknown> {
+interface ColumnValue<TRow, TDependentValue = any, TField extends keyof TRow = keyof TRow> {
   /** The name of the column. By default it will be displayed in the header cell */
   name: string;
   /** A unique key to distinguish each column */
-  key: keyof TRow;
+  key: TField;
```
Comment (amanmahajan7) [inline packages/react-data-grid/src/common/types.ts]: This makes value strongly typed

image

File: packages/react-data-grid/src/common/types.ts [hunk 4]
```diff
@@ -120,17 +124,17 @@ export interface Editor<TValue = never> extends React.Component {
   readonly disableContainerStyles?: boolean;
 }
 
-export interface FormatterProps<TValue, TDependentValue = unknown, TRow = any> {
+export interface FormatterProps<TValue, TDependentValue = any, TRow = any> {
```
Comment (amanmahajan7) [inline packages/react-data-grid/src/common/types.ts]: I put `any` so we can create generic formatters. `unknown` might work, I will have to check

File: packages/react-data-grid/src/Row.tsx [hunk 5]
```diff
@@ -62,7 +62,7 @@ export default class Row<R extends {}> extends React.Component<RowRendererProps<
       lastFrozenColumnIndex
     };
 
-    return <Renderer key={`${key}-${idx}`} {...cellProps} />;
+    return <Renderer key={`${key as keyof R}-${idx}`} {...cellProps} />; // FIXEME: fix key type
```
Comment (MayhemYDG) [inline packages/react-data-grid/src/Row.tsx]:
```suggestion
    return <Renderer key={`${key as keyof R}-${idx}`} {...cellProps} />; // FIXME: fix key type
```
Might miss this one when searching for `fixme`.

**Comment by amanmahajan7** [comment]
I think this change also highlights some of the issues with RDG, for example

```
value: this.getCellValue(key || String(column.idx));
```

There is no way to type this. I think we can just use `rowData[key]`

```
<SimpleCellFormatter value={value as string} />
```

We should probably change `SimpleCellFormatter` and change `value` from `string | number | boolean` -> `unknown` as we use this formatter for any column

Regarding keys, because we use `keyof R` type for `column.key` instead of a string so we get this type. I am not too sure how to fix it yet.

image. 

Using strongly type value, formatters, and editors did surfaced quite a few type issues in our code base, I think we can try these types and revisit for improvement.

**Pull request merged**